libraries/ESP8266MQTTClient/src/ESP8266MQTTClient.cpp

```diff
@@ -0,0 +1,522 @@
+    }
+
+    if (!_transportTraits) {
+        LOG("[MQTT-Client] connect: MQTTClient::begin was not called or returned error\n");
+        return false;
+    }
+
+    if(_scheme == "mqtt")
+        _tcp = _transportTraits->create();
+    else {
+        LOG("ERROR: currently only support mqtt over tcp\r\n");
+        return false;
+    }
+
+    if(!_tcp->connect(_host.c_str(), _port)) {
+        LOG("[MQTT-Client] failed connect to %s:%u\n", _host.c_str(), _port);
+        return false;
+    }
+
+    LOG("[MQTT-Client] connected to %s:%u\n", _host.c_str(), _port);
+
+    if (!_transportTraits->verify(*_tcp, _host.c_str())) {
+        LOG("[MQTT-Client] transport level verify failed\n");
+        _tcp->stop();
+        return false;
+    }
+
+    _tcp->setNoDelay(true);
+
+    _mqtt_state.outbound_message = mqtt_msg_connect(&_mqtt_state.mqtt_connection,
+                                   _mqtt_state.connect_info);
+    _mqtt_state.pending_msg_type = mqtt_get_type(_mqtt_state.outbound_message->data);
+    _mqtt_state.pending_msg_id = mqtt_get_id(_mqtt_state.outbound_message->data,
+                                 _mqtt_state.outbound_message->length);
+    LOG("Sending MQTT CONNECT message, type: %d, id: %04X\r\n",
+        _mqtt_state.pending_msg_type,
+        _mqtt_state.pending_msg_id);
+    write_len = _tcp->write(_mqtt_state.outbound_message->data,
+                            _mqtt_state.outbound_message->length);
+    LOG("Reading MQTT CONNECT response message\r\n");
+    while (_tcp->available() == 0);
+    read_len = _tcp->read(_mqtt_state.in_buffer, DEFAULT_MQTT_BUFFER_SIZE_BYTES);
+
+
+
+    if (read_len < 0) {
+        LOG("Error network response\r\n");
+        return false;
+    }
+    if (mqtt_get_type(_mqtt_state.in_buffer) != MQTT_MSG_TYPE_CONNACK) {
+        LOG("Invalid MSG_TYPE response: %d, read_len: %d\r\n", mqtt_get_type(_mqtt_state.in_buffer), read_len);
+        return false;
+    }
+    connect_rsp_code = mqtt_get_connect_return_code(_mqtt_state.in_buffer);
+    switch (connect_rsp_code) {
+        case CONNECTION_ACCEPTED:
+            LOG("Connected\r\n");
+            return connected();
+        case CONNECTION_REFUSE_PROTOCOL:
+        case CONNECTION_REFUSE_SERVER_UNAVAILABLE:
+        case CONNECTION_REFUSE_BAD_USERNAME:
+        case CONNECTION_REFUSE_NOT_AUTHORIZED:
+            LOG("Connection refuse, reason code: %d\r\n", connect_rsp_code);
+            return false;
+        default:
+            LOG("Connection refuse, Unknow reason\r\n");
+            return false;
+    }
+    return false;
+}
+
+void MQTTClient::onConnect(THandlerFunction fn)
+{
+    _connected_cb = fn;
+}
+void MQTTClient::onSubscribe(THandlerFunction_PubSub fn)
+{
+    _subscribe_cb = fn;
+}
+void MQTTClient::onPublish(THandlerFunction_PubSub fn)
+{
+    _publish_cb = fn;
+}
+void MQTTClient::onData(THandlerFunction_Data fn)
+{
+    _data_cb = fn;
+}
+void MQTTClient::handle(void)
+{
+    mqtt_outbox *ob;
+    while(!connected()) {
```

Comment:
this will block the whole sketch if connects fail. Maybe you want to return on connect fail and do not try to connect for ```MQTT_RECONNECT_TIMEOUT```, but let the rest of the sketch run.